### PR TITLE
UCT/UGNI: Start using spinlocks to protect critical structures

### PR DESCRIPTION
**Pull request by MattBBaker**
This PR is deceptively big. A lot of it is shuffling a few big functions and rearranging some data structures to allow a shared thread safe code path. I couldn't find a useful way of splitting it up more.

What it does is add a spin lock to the device, and that is what will get used to do locking in UGNI. This lock is then used to attach the CDM to the device. The cdm device fields are then merged together into a shared struct.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1607/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3528/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
@shamisp Could you review?

**Comment by shamisp** [comment]
v1.3, right ?

**Comment by MattBBaker** [comment]
@shamisp Yes, this a bit much of a change for 1.2.

**Comment by shamisp** [comment]
@MattBBaker it is actually up to you. I think these changes can be considered as a thread safety bugfix. If you think it is important for ORNL to have it in v1.2 we can push.
I don't think @yosefe will have major objection

**Comment by MattBBaker** [comment]
@shamisp This is part of a larger set of patches coming down the pipe , with some more rearrangements and potential performance impacts. I think that saying it's 1.3 is safer. Though I think that we should also talk about what kind of release time frame there is for 1.3.

**Comment by MattBBaker** [comment]
@shamisp Also, don't merge yet, I just found a bug.

**Comment by MattBBaker** [comment]
@shamisp Fixed now. I guess gcc doesn't check the parameter list for empty macros.

**Comment by shamisp** [comment]
it does not. Ready for review ?

**Comment by MattBBaker** [comment]
Yes.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1634/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3553/ for details (Mellanox internal link).

File: src/uct/ugni/smsg/ugni_smsg_ep.c [hunk 1]
```diff
@@ -53,7 +53,7 @@ static ucs_status_t uct_ugni_smsg_mbox_reg(uct_ugni_smsg_iface_t *iface, uct_ugn
     }
     pthread_mutex_lock(&uct_ugni_global_lock);
 
-    ugni_rc = GNI_MemRegister(iface->super.nic_handle, (uint64_t)address,
+    ugni_rc = GNI_MemRegister(uct_ugni_iface_nic_handle(&iface->super), (uint64_t)address,
```
Comment (shamisp) [inline src/uct/ugni/smsg/ugni_smsg_ep.c]: I'm curious - what is the motivation behind uct_ugni_iface_nic_handle() ? type more  :)

File: src/uct/ugni/udt/ugni_udt_iface.c [hunk 2]
```diff
@@ -479,8 +466,6 @@ static UCS_CLASS_INIT_FUNC(uct_ugni_udt_iface_t, uct_md_h md, uct_worker_h worke
         ucs_warn("GNI_EpDestroy failed, Error status: %s %d",
                  gni_err_str[ugni_rc], ugni_rc);
     }
- clean_iface_activate:
-    ugni_deactivate_iface(&self->super);
```
Comment (shamisp) [inline src/uct/ugni/udt/ugni_udt_iface.c]: Where did the cleanup go ?

File: src/uct/ugni/base/ugni_iface.h [hunk 3]
```diff
@@ -23,22 +23,24 @@ ucs_status_t uct_ugni_iface_get_address(uct_iface_h tl_iface, uct_iface_addr_t *
...
+    return iface->cdm.dev;
+}
+static inline gni_nic_handle_t uct_ugni_iface_nic_handle(uct_ugni_iface_t *iface)
+{
+    return iface->cdm.nic_handle;
+}
+static inline int uct_ugni_check_device_type(uct_ugni_iface_t *iface, gni_nic_device_t type)
+{
+    uct_ugni_device_t *dev = uct_ugni_iface_device(iface);
+    return dev->type == type;
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_iface.h]: bool instead of int ?
Also I can be a single line.

File: src/uct/ugni/base/ugni_iface.c [hunk 4]
```diff
@@ -350,20 +246,32 @@ UCS_CLASS_INIT_FUNC(uct_ugni_iface_t, uct_md_h md, uct_worker_h worker,
...
+    status = uct_ugni_create_cdm(&self->cdm, dev, worker->thread_mode);
+    if (UCS_OK != status) {
+        ucs_error("Failed to UGNI NIC, Error status: %d", status);
+        return status;
+    }
+    ugni_rc = GNI_CqCreate(uct_ugni_iface_nic_handle(self), UCT_UGNI_LOCAL_CQ, 0,
+                           GNI_CQ_NOBLOCK,
+                           NULL, NULL, &self->local_cq);
+    if (GNI_RC_SUCCESS != ugni_rc) {
+        ucs_error("GNI_CqCreate failed, Error status: %s %d",
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_iface.c]: Somebody has to clean CDM ?

File: src/uct/ugni/smsg/ugni_smsg_ep.c (hunk 1 above)
Comment (MattBBaker) [inline src/uct/ugni/smsg/ugni_smsg_ep.c]: It's most useful in UDT, where changes to the structure (like when I was deciding where that field should land in the final structure layout) required changes in a lot of place. This way, change it once and everything compiles fine. I can condense the name. :)

File: src/uct/ugni/udt/ugni_udt_iface.c (hunk 2 above)
Comment (MattBBaker) [inline src/uct/ugni/udt/ugni_udt_iface.c]: In a different commit that apparently I failed to include. Oops, will rectify that.

File: src/uct/ugni/smsg/ugni_smsg_ep.c (hunk 1 above)
Comment (shamisp) [inline src/uct/ugni/smsg/ugni_smsg_ep.c]: okay... you are the one that will be typing this

File: src/uct/ugni/udt/ugni_udt_iface.c (hunk 2 above)
Comment (shamisp) [inline src/uct/ugni/udt/ugni_udt_iface.c]: Overall I noticed that the clean up flow disappeared from a few places.

Comment (MattBBaker) [inline src/uct/ugni/udt/ugni_udt_iface.c]: Story is that there are multiple related commits, together they are massive so I broke them up. The clean up path was fixed in the next PR, but I need to fix it for this one.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1638/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1640/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3557/ for details (Mellanox internal link).

File: src/uct/ugni/udt/ugni_udt_iface.h [hunk 5]
```diff
@@ -101,4 +101,8 @@ static inline int uct_ugni_udt_ep_any_post(uct_ugni_udt_iface_t *iface)
     return UCS_OK;
 }
 
+static inline gni_nic_handle_t uct_ugni_udt_iface_nic_handle(uct_ugni_udt_iface_t *iface)
```
Comment (shamisp) [inline src/uct/ugni/udt/ugni_udt_iface.h]: Looks like a macro to me..

**Comment by shamisp** [comment]
@hppritcha if you have cycles, please take a look.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3559/ for details (Mellanox internal link).

File: src/uct/ugni/udt/ugni_udt_iface.h (hunk 5 above)
Comment (MattBBaker) [inline src/uct/ugni/udt/ugni_udt_iface.h]: Looks like an inline function. The one just pushed one looks like a macro.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1651/ for details.

**Comment by MattBBaker** [comment]
ORNL jenkins failure is an mlx failure that is already known. Not relevant to this PR.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3569/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
bot:ornl:retest

**Comment by MattBBaker** [comment]
@shamisp Updated. Should be good to merge now yes?

File: src/uct/ugni/base/ugni_iface.c [hunk 6]
```diff
@@ -378,18 +294,23 @@ UCS_CLASS_INIT_FUNC(uct_ugni_iface_t, uct_md_h md, uct_worker_h worker,
                             "UGNI-DESC-ONLY");
     if (UCS_OK != status) {
         ucs_error("Could not init iface");
+        goto clean_cq;
     }
     return status;
+clean_cq:
+    GNI_CqDestroy(self->local_cq);
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_iface.c]: Log error code

File: src/uct/ugni/smsg/ugni_smsg_iface.c [hunk 7]
```diff
@@ -217,12 +217,9 @@ static UCS_CLASS_CLEANUP_FUNC(uct_ugni_smsg_iface_t)
 {
     uct_worker_progress_unregister(self->super.super.worker,
                                    uct_ugni_smsg_progress, self);
-    if (!self->super.activated) {
-        return;
-    }
-
     ucs_mpool_cleanup(&self->free_desc, 1);
     ucs_mpool_cleanup(&self->free_mbox, 1);
+    GNI_CqDestroy(self->remote_cq);
```
Comment (shamisp) [inline src/uct/ugni/smsg/ugni_smsg_iface.c]: please log the error

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1693/ for details.

File: src/uct/ugni/base/ugni_iface.c [hunk 8]
```diff
@@ -234,136 +232,54 @@ static ucs_status_t uct_ugni_fetch_pmi()
...
     .chunk_release = ucs_mpool_chunk_free,
     .obj_init      = NULL,
     .obj_cleanup   = NULL
 };
 
+void uct_ugni_cleanup_base_iface(uct_ugni_iface_t *iface)
+{
+    ucs_arbiter_cleanup(&iface->arbiter);
+    ucs_mpool_cleanup(&iface->flush_pool, 1);
+    GNI_CqDestroy(iface->local_cq);
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_iface.c]: error log

File: src/uct/ugni/smsg/ugni_smsg_ep.c [hunk 9]
```diff
@@ -180,7 +180,7 @@ ucs_status_t uct_ugni_smsg_ep_connect_to_ep(uct_ep_h tl_ep,
     }
 
     ep_hash = (uint32_t)iface_addr->ep_hash;
-    gni_rc = GNI_EpSetEventData(ep->super.ep, iface->domain_id, ep_hash);
+    gni_rc = GNI_EpSetEventData(ep->super.ep, iface->cdm.domain_id, ep_hash);
```
Comment (shamisp) [inline src/uct/ugni/smsg/ugni_smsg_ep.c]: seems in some places we use gni_rc and another ugni_rc. IMHO should be consistent. (separate PR)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3623/ for details (Mellanox internal link).

File: src/uct/ugni/smsg/ugni_smsg_ep.c (hunk 9 above)
Comment (MattBBaker) [inline src/uct/ugni/smsg/ugni_smsg_ep.c]: Sounds like a plan.

File: src/uct/ugni/base/ugni_iface.c (hunk 8 above)
Comment (MattBBaker) [inline src/uct/ugni/base/ugni_iface.c]: @shamisp Thing is, the next PR pulls out all of the GNI_Cq* calls and has common spin locked code paths with error logging. I'd like to fix these problems in the next PR instead.

**Comment by shamisp** [comment]
Opened Issue #1529

**Pull request merged**